app/lib/oidc_client.rb

```diff
@@ -73,26 +75,50 @@ def set_checker_attribute(value:, access_token:, refresh_token:)
       access_token: access_token,
       refresh_token: refresh_token,
       method: :put,
+      uri: attribute_uri,
       arg: { value: value.to_json },
     )
   end
 
+  def has_email_subscription(access_token:, refresh_token:)
+    response = oauth_request(
+      access_token: access_token,
+      refresh_token: refresh_token,
+      method: :get,
+      uri: email_subscription_uri,
+    )
+
+    response.merge(result: (200..299).include?(response[:result].status))
+  end
+
+  def update_email_subscription(slug:, access_token:, refresh_token:)
+    oauth_request(
+      access_token: access_token,
+      refresh_token: refresh_token,
+      method: :post,
+      uri: email_subscription_uri,
+      arg: { topic_slug: slug },
+    )
+  end
+
 private
 
-  def oauth_request(access_token:, refresh_token:, method:, arg: nil)
+  OK_STATUSES = [200, 204, 404, 410].freeze
+
+  def oauth_request(access_token:, refresh_token:, method:, uri:, arg: nil)
     access_token_str = access_token
     refresh_token_str = refresh_token
 
-    args = [attribute_uri, arg].compact
+    args = [uri, arg].compact
 
     response = Rack::OAuth2::AccessToken::Bearer.new(access_token: access_token_str).public_send(method, *args)
 
-    unless [200, 404].include? response.status
+    unless OK_STATUSES.include? response.status
```

Comment:
Not sure i understand this bit.
So if our response isn't 200, 204, 404, 410 then we refresh...

Reply:
Would it help if I changed this to something like

```
if [401, 403].include? response.status
    # refresh tokens and retry
elsif response.status >= 500
    # retry
elsif OK_STATUSES.include? response.status
    # this is ok
else
    # this is an unrecoverable error
```